p2p/service.py

```diff
@@ -27,18 +26,35 @@ def __init__(self, token: CancelToken=None) -> None:
         else:
             self.cancel_token = base_token.chain(token)
 
+    async def wait_first(
+            self, *futures: Awaitable, token: CancelToken = None, timeout: float = None) -> Any:
+        """Wait for the first future to complete, unless we timeout or the token chain is triggered.
+
+        The given token is chained with this service's token, so triggering either will cancel
+        this.
+
+        Returns the result of the first future to complete.
+
+        Raises TimeoutError if we timeout or OperationCancelled if the token chain is triggered.
+
+        All pending futures are cancelled before returning.
+        """
+        if token is None:
+            token_chain = self.cancel_token
+        else:
+            token_chain = token.chain(self.cancel_token)
+        return await wait_with_token(*futures, token=token_chain, timeout=timeout)
+
     async def run(
             self,
             finished_callback: Optional[Callable[['BaseService'], None]] = None) -> None:
         """Await for the service's _run() coroutine.
 
-        Once _run() returns, set the finished event, call cleanup() and
+        Once _run() returns, triggers the cancel token, call cleanup() and
         finished_callback (if one was passed).
         """
         if self.is_running:
             raise RuntimeError("Cannot start the service while it's already running")
-        elif self.is_finished:
-            raise RuntimeError("Cannot restart a service after it has completed")
```

Comment:
This check has come in handy at least once. Instead of removing it, how about:

```py
        elif self.cancel_token.triggered:
            raise RuntimeError("Cannot restart a service after it has been canceled")
```

Reply:
Sure, I'll include that in one branch I'm currently working on